File: apps/web/src/lib/roles.ts
export type WorkspaceRole = "Admin" | "Manager" | "User";

export type AdminPermission =
  | "admin.access"
  | "workspace.read"
  | "workspace.write"
  | "users.read"
  | "users.invite"
  | "users.remove"
  | "users.role"
  | "knowledge.read"
  | "knowledge.write"
  | "ai.read"
  | "ai.write"
  | "usage.read"
  | "privacy.read"
  | "privacy.write"
  | "retention.write"
  | "audit.read"
  | "devices.read"
  | "devices.write"
  | "licenses.read"
  | "licenses.write";

const MATRIX: Record<WorkspaceRole, AdminPermission[]> = {
  Admin: [
    "admin.access",
    "workspace.read",
    "workspace.write",
    "users.read",
    "users.invite",
    "users.remove",
    "users.role",
    "knowledge.read",
    "knowledge.write",
    "ai.read",
    "ai.write",
    "usage.read",
    "privacy.read",
    "privacy.write",
    "retention.write",
    "audit.read",
    "devices.read",
    "devices.write",
    "licenses.read",
    "licenses.write",
  ],
  Manager: [
    "admin.access",
    "workspace.read",
    "users.read",
    "knowledge.read",
    "ai.read",
    "usage.read",
    "privacy.read",
    "audit.read",
    "devices.read",
    "licenses.read",
  ],
  User: [],
};

export function normalizeRole(value: unknown): WorkspaceRole {
  if (typeof value !== "string") return "User";
  const compact = value.trim();
  if (compact === "Admin" || compact === "Manager" || compact === "User") return compact;
  const v = compact.toLowerCase();
  if (v === "admin") return "Admin";
  if (v === "manager") return "Manager";
  if (v === "user" || v === "member" || v === "viewer") return "User";
  return "User";
}

export function permissionsFor(role?: string | null): AdminPermission[] {
  return MATRIX[normalizeRole(role)] || [];
}

export function rolePermissionMatrix(): Record<WorkspaceRole, AdminPermission[]> {
  return {
    Admin: [...MATRIX.Admin],
    Manager: [...MATRIX.Manager],
    User: [...MATRIX.User],
  };
}

export const ALL_ADMIN_PERMISSIONS: AdminPermission[] = [
  "admin.access",
  "workspace.read",
  "workspace.write",
  "users.read",
  "users.invite",
  "users.remove",
  "users.role",
  "knowledge.read",
  "knowledge.write",
  "ai.read",
  "ai.write",
  "usage.read",
  "privacy.read",
  "privacy.write",
  "retention.write",
  "audit.read",
  "devices.read",
  "devices.write",
  "licenses.read",
  "licenses.write",
];

export function can(role: string | null | undefined, permission: AdminPermission): boolean {
  return permissionsFor(role).includes(permission);
}

export function canAccessAdmin(role?: string | null): boolean {
  return can(role, "admin.access");
}

File: apps/api/app/core/roles.py
from __future__ import annotations

from typing import Literal

WorkspaceRole = Literal["Admin", "Manager", "User"]
AdminPermission = str

_MATRIX: dict[WorkspaceRole, list[AdminPermission]] = {
    "Admin": [
        "admin.access",
        "workspace.read",
        "workspace.write",
        "users.read",
        "users.invite",
        "users.remove",
        "users.role",
        "knowledge.read",
        "knowledge.write",
        "ai.read",
        "ai.write",
        "usage.read",
        "privacy.read",
        "privacy.write",
        "retention.write",
        "audit.read",
        "devices.read",
        "devices.write",
        "licenses.read",
        "licenses.write",
    ],
    "Manager": [
        "admin.access",
        "workspace.read",
        "users.read",
        "knowledge.read",
        "ai.read",
        "usage.read",
        "privacy.read",
        "audit.read",
        "devices.read",
        "licenses.read",
    ],
    "User": [],
}


def normalize_role(value: object | None) -> WorkspaceRole:
    if not isinstance(value, str):
        return "User"
    compact = value.strip()
    if compact in ("Admin", "Manager", "User"):
        return compact  # type: ignore[return-value]
    lowered = compact.lower()
    if lowered == "admin":
        return "Admin"
    if lowered == "manager":
        return "Manager"
    if lowered in {"user", "member", "viewer"}:
        return "User"
    return "User"


def permissions_for(role: str | None) -> list[AdminPermission]:
    return list(_MATRIX.get(normalize_role(role), []))


def can(role: str | None, permission: AdminPermission) -> bool:
    return permission in permissions_for(role)


def can_access_admin(role: str | None) -> bool:
    return can(role, "admin.access")
